File: R/gaussianMixture-internals.R
#' Moments of a Gaussian Mixture
#'
#' Compute the first four moments, central moments, mean, variance, skewness and kurtosis
#' for a Gaussian Mixture with K components.
#'
#' @examples
#' means = c(-0.9, 0.5)
#' sd = c(0.4,1)
#' alpha = c(0.5, 0.5)
#' GM_moments(means, sd, alpha)
#' GM_moments(c(-0.8, 1.8), c(0.4,1), c(0.5, 0.5))
#'
#' @details
#' The non-central moments are denoted as `m1`, `m2`, `m3`, `m4`. The central moments as `mu2`, `mu3`, `mu4`.
#' @rdname GM_moments
#' @name GM_moments
#' @references https://en.wikipedia.org/wiki/Normal_distribution#Moments
#' @keywords gaussianMixture
#' @note Version 1.0.0
#' @export
#' @noRd
GM_moments <- function(means, sd, alpha){
  # First moment (Expectation)
  m1 <- sum(means * alpha)
  # Non-central moments
  # Second moment
  m2 <- sum((means^2 + sd^2) * alpha)
  # Third moment
  m3 <- sum((3 * means * sd^2 + means^3) * alpha)
  # Fourth moment
  m4 <- sum((3 * sd^4 + 6 * means^2 * sd^2 + means^4) * alpha)
  # Fifth moment
  # m5 <- sum((means^5 + 10 * means^3 * sd^2 + 15 * means * sd^4) * alpha)
  # Sixth moment
  # m6 <- sum((means^6 + 15 * means^4 * sd^2 + 45 * means^2 * sd^4 + 15 * sd^6)*alpha)
  # Central-moments
  delta_k <- means - m1
  # Second moment
  mu2 <- sum((delta_k^2 + sd^2) * alpha)
  # Third moment
  mu3 <- sum((3 * delta_k * sd^2 + delta_k^3) * alpha)
  # Fourth moment
  mu4 <- sum((3 * sd^4 + 6 * delta_k^2 * sd^2 + delta_k^4) * alpha)
  # Fifth moment
  # mu5 <- sum((delta_k^5 + 10 * delta_k^3 * sd^2 + 15 * delta_k * sd^4) * alpha)
  # Sixth moment
  # mu6 <- sum((delta_k^6 + 15 * delta_k^4 * sd^2 + 45 * delta_k^2 * sd^4 + 15 * sd^6)*alpha)

  # Derived statistics
  # Variance
  v_x <- mu2
  # Skewness
  sk_x <- mu3 / (v_x^(3/2))
  # Excess Kurtosis
  kt_x <- mu4 / (v_x^2) - 3

  dplyr::tibble(
    m1 = m1,
    m2 = m2,
    m3 = m3,
    m4 = m4,
    mu2 = mu2,
    mu3 = mu3,
    mu4 = mu4,
    mean = m1,
    variance = v_x,
    skewness = sk_x,
    kurtosis = kt_x
  )
}

#' Match the first three moments of a Gaussian Mixture
#'
#' @param d Numeric, distance between the two means.
#' @param m1 Numeric, first target moment.
#' @param m2 Numeric, second target moment.
#' @param m3 Numeric, third target moment.
#' @param p Numeric, probability.
#'
#' @keywords gaussianMixture
#' @note Version 1.0.0
#' @export
#' @noRd
GM_moments_match <- function(d, m1 = 0, m2 = 1, m3 = 0, p = 0.5){
  means <- c(mu1 = 0, mu2 = 0)
  sigma <- c(sd1 = 0, sd2 = 0)
  probs <- c(p1 = p, p2 = 1-p)
  # First moment
  means[1] <- m1 + (1-p) * d
  means[2] <- m1 - p * d
  # Second moment
  delta <- (m3 - p * (1 - p) * ((1 - p)^2 - p^2) * d^3) / (3 * p * (1-p) * d)
  sigma[2] <- m2 - p * (1 - p) * d^2 - p * delta
  sigma[1] <- sigma[2] + delta
  sigma <- sqrt(sigma)

  structure(
    list(
      means = means,
      sigma = sigma,
      probs = probs
    )
  )
}

#' Compute the log-likelihood of a Gaussian Mixture
#'
#' @param means description
#' @param sd description
#' @param alpha description
#' @examples
#' GM_loglik(c(-0.8, 0.8), c(0.4,1), c(0.5, 0.5), rnorm(100))
#'
#' @keywords gaussianMixture
#' @note Version 1.0.0
#' @export
#' @noRd
GM_loglik <- function(means, sd, alpha, x){
  # Log-likelihood
  if (!missing(x)) {
    lik <- dmixnorm(x, means, sd, alpha)
    loss <- sum(log(lik))
  } else {
    loss <- 0
  }
  return(loss)
}

#' Fit a Gaussian Mixture model
#'
#' @param x Numeric, vector on which the model will be fitted.
#' @param method Character, method use to fit the GM model, can be `mclust` or `mixtools`.
#' @param components Integer, number of components for the mixture.
#'
#' @examples
#' x <- rmixnorm(1000, c(0, 0), c(0.3, 2), c(0.4, 0.6))
#' # Fitted parameters without contraints
#' params <- GM_fit(x$X)
#' GM_moments(params$means, params$sd, params$p)
#'
#' @keywords gaussianMixture
#' @note Version 1.0.0
#' @export
#' @noRd
GM_fit <- function(x, method = c("mclust", "mixtools"), components = 2, maxit = 30000, maxrestarts = 500){
  clust <- NULL
  method <- match.arg(method, choices = c("mclust", "mixtools"))
  # Fitted parameters with mclust
  if (method == "mclust") {
    clust <- mclust::Mclust(x, G = components, modelNames = c("V"), verbose = FALSE)
    if (!is.null(clust)){
      # Estimated parameters
      means <- clust$parameters$mean
      sd <- sqrt(clust$parameters$variance$scale)
      p <- clust$parameters$pro
    }
  }
  # Fitted parameters with mixtools
  if (method == "mixtools" | is.null(clust)) {
    quiet_EM <- purrr::quietly(mixtools::normalmixEM)
    clust <- quiet_EM(x, maxit = maxit, k = components, maxrestarts = maxrestarts)$result
    # Estimated parameters
    means <- clust$mu
    sd <- clust$sigma
    p <- clust$lambda
  }
  idx <- order(means)
  means = means[idx]
  sd = sd[idx]
  p = p[idx]
  # Assign a name to the parameters
  names(means) <- paste0("mu", 1:components)
  names(sd) <- paste0("sd", 1:components)
  names(p) <- paste0("p", 1:components)

  structure(
    list(
      means = means,
      sd = sd,
      p = p
    )
  )
}

#' Fit a Gaussian Mixture with two components under moments constraints.
#'
#' @param x Numeric, vector on which the model will be fitted.
#' @param start_params Numeric vector of parameters
#' @param mu_target Numeric, scalar. Target mean of the mixture. If missing will be not applied any contraint.
#' @param var_target Numeric, scalar. Target variance of the mixture. If missing will be not applied any contraint.
#'
#' @examples
#' x <- rmixnorm(1000, c(0, 0), c(0.3, 2), c(0.4, 0.6))
#' # Fitted parameters without contraints
#' params <- GM_fit(x$X)
#' GM_moments(params$means, params$sd, params$p)
#'
#' # Fit the parameters with moments contraints
#' start_params <- unlist(purrr::flatten(params)[-6])
#' # Match a certain mean
#' match_params <- GM_fit_moments_match(x$X, start_params, mu_target = -0.09)
#' GM_moments(match_params$means, match_params$sd, match_params$p)
#' # Match a certain variance
#' match_params <- GM_fit_moments_match(x$X, start_params, var_target = 2.4)
#' GM_moments(match_params$means, match_params$sd, match_params$p)
#'
#' # Match a certain mean and variance
#' match_params <- GM_fit_moments_match(x$X, start_params, mu_target = -0.09, var_target = 2.4)
#' GM_moments(match_params$means, match_params$sd, match_params$p)
#'
#' @keywords gaussianMixture
#' @note Version 1.0.0
#' @export
#' @noRd
GM_fit_moments_match <- function(x, start_params, mu_target = NA, var_target = NA, maxit = 1000, abstol = 1e-5){
  # Objective: weighted log-likelihood
  objective <- function(par){
    mu1 <- par[1]
    mu2 <- par[2]
    sd1 <- par[3]
    sd2 <- par[4]
    p1 <- par[5]
    p2 <- 1 - par[5]
    # Log-likelihood
    loglik <- sum(log(p1 * dnorm(x, mu1, sd1) + p2 * dnorm(x, mu2, sd2)), na.rm = TRUE)
    return(-loglik)
  }

  # Equality constraints: match mean and variance
  constraints <- function(par){
    mu1 <- par[1]
    mu2 <- par[2]
    sd1 <- par[3]
    sd2 <- par[4]
    p1 <- par[5]
    p2 <- 1 - par[5]
    # Moments
    e_x <- p1 * mu1 + p2 * mu2
    e_x2 <- p1 * (sd1^2 + mu1^2) + p2 * (sd2^2 + mu2^2)
    v_x <- e_x2 - e_x^2

    if (is.na(mu_target) & !is.na(var_target)){
      loss <- c(0, v_x - var_target)
    } else if (!is.na(mu_target) & is.na(var_target)){
      loss <- c(e_x - mu_target, 0)
    } else {
      loss <- c(e_x - mu_target, v_x - var_target)
    }
    return(loss)
  }
  # Solver with constraints
  res <- nloptr::nloptr(
    x0 = start_params,
    eval_f = objective,
    eval_g_eq = constraints,
    # Lower bounds
    lb = c(-3, -3, 0.01, 0.01, 0.01),
    # Upper bounds
    ub = c(3, 3, 3, 3, 0.99),
    opts = list(algorithm = "NLOPT_LN_COBYLA", xtol_rel = abstol, maxeval = maxit)
  )
  structure(
    list(
      means = res$solution[1:2],
      sd = res$solution[3:4],
      p = c(res$solution[5], 1-res$solution[5])
    )
  )
}

File: solarrpy/gaussianMixture_internals.py
import numpy as np
import pandas as pd
from scipy.stats import norm
from typing import Dict
from sklearn.mixture import GaussianMixture
from scipy.optimize import minimize
from typing import Dict, Optional

def gm_moments(means: np.ndarray, sd: np.ndarray, alpha: np.ndarray) -> pd.DataFrame:
    """
    Compute the first four non-central and central moments, mean, variance, 
    skewness, and excess kurtosis for a Gaussian Mixture.
    """
    # Ensure inputs are numpy arrays for vectorized operations
    means = np.asarray(means, dtype=float)
    sd = np.asarray(sd, dtype=float)
    alpha = np.asarray(alpha, dtype=float)
    
    # First moment (Expectation)
    m1 = np.sum(means * alpha)
    
    # Non-central moments
    m2 = np.sum((means**2 + sd**2) * alpha)
    m3 = np.sum((3 * means * sd**2 + means**3) * alpha)
    m4 = np.sum((3 * sd**4 + 6 * means**2 * sd**2 + means**4) * alpha)
    
    # Central moments: shift means by the global expectation
    delta_k = means - m1
    mu2 = np.sum((delta_k**2 + sd**2) * alpha)
    mu3 = np.sum((3 * delta_k * sd**2 + delta_k**3) * alpha)
    mu4 = np.sum((3 * sd**4 + 6 * delta_k**2 * sd**2 + delta_k**4) * alpha)
    
    # Derived statistics
    v_x = mu2
    sk_x = mu3 / (v_x**(1.5))
    kt_x = (mu4 / (v_x**2)) - 3.0

    # Return as a 1-row DataFrame to mimic dplyr::tibble behavior
    return pd.DataFrame({
        'm1': [m1],
        'm2': [m2],
        'm3': [m3],
        'm4': [m4],
        'mu2': [mu2],
        'mu3': [mu3],
        'mu4': [mu4],
        'mean': [m1],
        'variance': [v_x],
        'skewness': [sk_x],
        'kurtosis': [kt_x]
    })

def gm_moments_match(d: float, m1: float = 0.0, m2: float = 1.0, m3: float = 0.0, p: float = 0.5) -> Dict[str, np.ndarray]:
    """
    Match the first three moments of a 2-component Gaussian Mixture.
    """
    # Mixing probabilities
    probs = np.array([p, 1 - p], dtype=float)
    
    # 1. Solve for means using the first moment and distance 'd'
    mu1 = m1 + (1 - p) * d
    mu2 = m1 - p * d
    means = np.array([mu1, mu2], dtype=float)
    
    # 2. Solve for variances using the second and third moments
    # delta represents the difference in variances: var1 - var2
    delta = (m3 - p * (1 - p) * ((1 - p)**2 - p**2) * d**3) / (3 * p * (1 - p) * d)
    
    var2 = m2 - p * (1 - p) * d**2 - p * delta
    var1 = var2 + delta
    
    # 3. Convert variances to standard deviations
    sigma = np.sqrt(np.array([var1, var2], dtype=float))
    
    return {
        'means': means,
        'sigma': sigma,
        'probs': probs
    }

def gm_loglik(means: np.ndarray, sd: np.ndarray, alpha: np.ndarray, x: np.ndarray = None) -> float:
    """
    Compute the log-likelihood of a Gaussian Mixture.
    """
    if x is not None:
        # Ensure inputs are numpy arrays
        x = np.asarray(x, dtype=float)
        means = np.asarray(means, dtype=float)
        sd = np.asarray(sd, dtype=float)
        alpha = np.asarray(alpha, dtype=float)
        
        # Calculate the density for each component for each observation.
        # x[:, np.newaxis] reshapes x to an (N, 1) column vector.
        # means, sd, and alpha are (K,) row vectors.
        # Broadcasting creates an (N, K) matrix of component likelihoods.
        component_densities = norm.pdf(x[:, np.newaxis], loc=means, scale=sd)
        
        # Multiply by mixture weights and sum across components (axis=1) 
        # to get the total mixture density per observation (N,)
        mix_density = np.sum(alpha * component_densities, axis=1)
        
        # Sum the log of the densities to get the total log-likelihood
        loss = np.sum(np.log(mix_density))
    else:
        loss = 0.0
        
    return loss

def gm_fit(x: np.ndarray, components: int = 2, maxit: int = 30000, maxrestarts: int = 10) -> Dict[str, np.ndarray]:
    """
    Fit a Gaussian Mixture model using scikit-learn's Expectation-Maximization.
    """
    # scikit-learn expects 2D arrays of shape (n_samples, n_features)
    # We force the 1D input into a 2D column vector.
    x_matrix = np.asarray(x, dtype=float).reshape(-1, 1)
    
    # Initialize and fit the standard EM algorithm.
    # 'full' covariance type allows each component to have its own variance,
    # mapping exactly to mclust's modelNames = "V".
    gmm = GaussianMixture(
        n_components=components,
        covariance_type='full',
        max_iter=maxit,
        n_init=maxrestarts, # Number of initializations to run
        init_params='kmeans' # Standard robust initialization
    )
    
    gmm.fit(x_matrix)

    if not gmm.converged_:
        raise ValueError("Gaussian Mixture did not converge!")
    
    # Extract parameters. 
    # gmm.means_ is shape (n_components, 1) -> flatten to 1D
    # gmm.covariances_ is shape (n_components, 1, 1) for 1D 'full' -> flatten and sqrt
    # gmm.weights_ is shape (n_components,)
    means = gmm.means_.flatten()
    sd = np.sqrt(gmm.covariances_.flatten())
    p = gmm.weights_.flatten()
    
    # Enforce identifiability: Sort components by mean (Label Switching Fix)
    idx = np.argsort(means)
    
    return {
        'means': means[idx],
        'sd': sd[idx],
        'p': p[idx]
    }

def gm_fit_moments_match(x: np.ndarray, 
                         start_params: np.ndarray, 
                         mu_target: Optional[float] = None, 
                         var_target: Optional[float] = None, 
                         maxit: int = 1000, 
                         abstol: float = 1e-5) -> Dict[str, np.ndarray]:
    """
    Fit a 2-component Gaussian Mixture matching target mean and/or variance.
    """
    # Pre-process data: drop NaNs once to avoid calculating na.rm=True on every iteration
    x = np.asarray(x, dtype=float)
    x = x[~np.isnan(x)]
    
    def objective(par):
        mu1, mu2, sd1, sd2, p1 = par
        p2 = 1.0 - p1
        
        # Calculate likelihoods
        density = p1 * norm.pdf(x, loc=mu1, scale=sd1) + p2 * norm.pdf(x, loc=mu2, scale=sd2)
        
        # Prevent log(0) underflow
        density = np.clip(density, a_min=1e-300, a_max=None)
        
        loglik = np.sum(np.log(density))
        return -loglik

    def constraints(par):
        mu1, mu2, sd1, sd2, p1 = par
        p2 = 1.0 - p1
        
        # Moments
        e_x = p1 * mu1 + p2 * mu2
        e_x2 = p1 * (sd1**2 + mu1**2) + p2 * (sd2**2 + mu2**2)
        v_x = e_x2 - e_x**2
        
        loss = []
        if mu_target is not None:
            loss.append(e_x - mu_target)
        if var_target is not None:
            loss.append(v_x - var_target)
            
        return np.array(loss)

    # Hardcoded bounds from the R implementation
    bounds = [
        (-3.0, 3.0),   # mu1
        (-3.0, 3.0),   # mu2
        (0.01, 3.0),   # sd1
        (0.01, 3.0),   # sd2
        (0.01, 0.99)   # p1
    ]
    
    # Define SciPy equality constraints
    cons = []
    if mu_target is not None or var_target is not None:
        cons.append({'type': 'eq', 'fun': constraints})
        
    # Run SLSQP Optimizer
    res = minimize(
        fun=objective,
        x0=start_params,
        method='SLSQP',
        bounds=bounds,
        constraints=cons,
        options={'maxiter': maxit, 'ftol': abstol}
    )
    
    if not res.success:
        # Note: Depending on pipeline architecture, you might want to raise an Exception here
        print(f"Optimization Warning: {res.message}")

    return {
        'means': res.x[0:2],
        'sd': res.x[2:4],
        'p': np.array([res.x[4], 1.0 - res.x[4]])
    }